### PR TITLE
Feature/zen 18191

### PR DESCRIPTION
**Pull request by dbouchillon**
changes to make the lpu audit script work better by using regexes.
changed lpu script so that we do not add user to namespaces more than once and do not show misleading error when setting dcom permissions
added example backup script

### CODE REVIEW

File: lpu/zenoss-audit-lpu.ps1 [hunk 1]
```diff
@@ -137,21 +138,29 @@ function is_user_in_group($groupname) {
...
 
 function is_user_in_service($service){
-    $testusersddl = "(A;;CCLCRPLORC;;;{0})" -f $usersid
+    $testuserperms = @('CC','LC','RP','LO','RC')
 	$servicesddlstart = [string](CMD /C "sc sdshow `"$service`"")
-	return $servicesddlstart.contains($testusersddl)
+    $intersection = @('place','holder')
+    foreach ($sddlkey in $servicesddlstart.split('('))  {
+        if ($sddlkey -match $usersid) {
+            $perms = [regex]::match($sddlkey, '\w;;(\w+);.*').Groups[1].value
```
Comment (cajoz) [inline lpu/zenoss-audit-lpu.ps1]: Does .NET/PowerShell cache regular expression objects? If not, might want to compile outside the loop.

File: lpu/zenoss-backup-lpu.ps1 [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+#
+# It does not determine WMI Namespace or Local Group information.
+#
+# Example usage:  PS > .\zenoss-backup-lpu.ps1 > backup.txt
+
+
+function backup_winrm_access(){
+    $sddlkey = "HKLM:SOFTWARE\Microsoft\Windows\CurrentVersion\WSMAN\Service"
+    $rootsddlkey = get-itemproperty $sddlkey -Name “rootSDDL”
+    Write-Output "winrm:  "$rootsddlkey.rootSDDL
```
Comment (cajoz) [inline lpu/zenoss-backup-lpu.ps1]: String formatting might be more readable.

File: lpu/zenoss-backup-lpu.ps1 [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+function backup_registry_sddl($regkey){
+    $regaclitem = get-item $regkey -ea silentlycontinue
+    if ($regaclitem -ne $null) {
+        $sddl = $regaclitem.getaccesscontrol("Access").sddl
+        Write-Output $regkey":  "$sddl
```
Comment (cajoz) [inline lpu/zenoss-backup-lpu.ps1]: Ditto.

File: lpu/zenoss-backup-lpu.ps1 [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    }
+}
+
+function backup_dcom_permissions($regkey, $property) {
+    $objSDHelper = New-Object System.Management.ManagementClass Win32_SecurityDescriptorHelper
+    $objRegProperty = Get-ItemProperty $regkey -Name $property -ea SilentlyContinue
+    if ($objRegProperty -ne $null) {
+        $sddl = [string]($objSDHelper.BinarySDToSDDL($objRegProperty.$property)).SDDL
+        Write-Output $regkey":"$property":  "$sddl
```
Comment (cajoz) [inline lpu/zenoss-backup-lpu.ps1]: Ditto.

File: lpu/zenoss-backup-lpu.ps1 [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    }
+}
+
+function backup_folderfile($folderfile) {
+    $sddl = (get-item $folderfile).getaccesscontrol("Access").Sddl
+    Write-Output $folderfile":  "$sddl
```
Comment (cajoz) [inline lpu/zenoss-backup-lpu.ps1]: Ditto.

File: lpu/zenoss-backup-lpu.ps1 [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+function backup_service_sddl($service) {
+    $servicesddl = [string](CMD /C "sc sdshow `"$service`"")
+    Write-Output "{$service}:  "$servicesddl
```
Comment (cajoz) [inline lpu/zenoss-backup-lpu.ps1]: Ditto.

File: lpu/zenoss-backup-lpu.ps1 [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+###################################################
+# The least privileged user needs winrm access
+# Write out the current access
+###################################################
+Write-Output("==== WinRM Access ====")
```
Comment (cajoz) [inline lpu/zenoss-backup-lpu.ps1]: Are parentheses required?

File: lpu/zenoss-backup-lpu.ps1 [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+backup_winrm_access
+
+####################################################
+# The lpu needs registry key access
+# Write out the current access to the necessary keys
+####################################################
+Write-Output("`n`n==== Registry Keys Permissions ====")
```
Comment (cajoz) [inline lpu/zenoss-backup-lpu.ps1]: Ditto.

File: lpu/zenoss-backup-lpu.ps1 [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+foreach ($registrykey in $registrykeys) {
+    backup_registry_sddl $registrykey
+}
+
+###################################################
+# The least privileged user needs DCOM access
+# Write out the current access for each registry value
+# DefaultAccessPermission may not exist, which is OK
+###################################################
+Write-Output("`n`n==== DCOM Permissions ====")
```
Comment (cajoz) [inline lpu/zenoss-backup-lpu.ps1]: Ditto.

File: lpu/zenoss-backup-lpu.ps1 [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+
+foreach ($registryvaluekey in $registryvaluekeys.GetEnumerator()){
+    backup_dcom_permissions $registryvaluekey.Value $registryvaluekey.Name
+}
+
+###################################################
+# The least privileged user needs folder access
+# Write out the current access for each folder
+###################################################
+Write-Output("`n`n==== Folder Permissions ====")
```
Comment (cajoz) [inline lpu/zenoss-backup-lpu.ps1]: Ditto.

File: lpu/zenoss-backup-lpu.ps1 [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+foreach($folderfile in $folderfiles){
+	backup_folderfile $folderfile 
+}
+
+
+###################################################
+# The least privileged user needs query access to services
+# Write out the current access for each service
+###################################################
+Write-Output("`n`n==== Service Permissions ====")
```
Comment (cajoz) [inline lpu/zenoss-backup-lpu.ps1]: Ditto.

**Comment by dbouchillon** [comment]
cajoz signed off on new changes

**Pull request merged**